packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -3792,12 +3792,9 @@ function commitDoubleInvokeEffectsInDEV(
   hasPassiveEffects: boolean,
 ) {
   if (__DEV__) {
-    if (useModernStrictMode) {
+    if (useModernStrictMode && root.tag !== LegacyRoot) {
       let doubleInvokeEffects = true;
 
-      if (root.tag === LegacyRoot && !(root.current.mode & StrictLegacyMode)) {
-        doubleInvokeEffects = false;
-      }
```

Comment:
@sammy-SC @acdlite, I fixed this bug that turned on strict effects in Legacy Mode, which may be the reason it was hard to land internally, breaking unit tests and whatnot.

Previously, `legacyCommitDoubleInvokeEffectsInDEV` would be called in the legacy root too (when strict mode was on), but we wouldn't fire any effects because no fiber flags were set. When we copied over the check to this block, it forces effects to be fired, since it's not checking fiber flags (by design).

We have a test that caught this, but we didn't have any test variants that enabled `useModernStrictMode`: 

https://github.com/facebook/react/blob/30e2938e04c8cf51688509a457a494d36bcc4269/packages/react-reconciler/src/__tests__/StrictEffectsMode-test.js#L29